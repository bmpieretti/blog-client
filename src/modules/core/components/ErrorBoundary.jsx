import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BoundaryError from '../styled/ErrorBoundary';
import { ErrorP } from '../../../styles/Base';

export default class ErrorBoundary extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ])
  }

  static defaultProps = {
    children: null
  }

  state = { error: '', infoStack: '' }

  componentDidCatch(error, info) {
    this.setState(prevState => ({
      ...prevState,
      error: error.message,
      infoStack: info.componentStack
    }));
  }

  render() {
    const { error, infoStack } = this.state;

    if (!error && !infoStack) {
      return this.props.children;
    }

    return (
      <BoundaryError>
        <h1>
          Oh no! You have hit an error that should never have made it into the app!
        </h1>
        <ErrorP>
          {error}
        </ErrorP>
        <footer>
          Let me know about this problem via email at <a href={`mailto:bmpieretti@gmail.com?subject=Incident%20Report&amp;body=${error}--${infoStack}`}>bmpieretti@gmail.com</a>.
        </footer>
      </BoundaryError>
    );
  }
}
