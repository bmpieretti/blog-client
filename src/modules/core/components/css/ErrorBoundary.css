.boundary-error {
  max-width: 480px;
  margin: 0 auto;
  align-self: center;
  composes: card from './Card.css';
}

.boundary-error-title {
  composes: card-h1 from './Card.css';
  text-align: center;
}

.boundary-error-message {
  composes: red-color from './Color.css';
  margin-bottom: 20px;
  text-align: center;
}

.boundary-error-footer {
  text-align: center;
}
