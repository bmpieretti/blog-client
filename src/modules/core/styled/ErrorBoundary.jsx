import styled from 'styled-components';

export default styled.div`
  max-width: 480px;
  margin: 0 auto;
  align-self: center;
  padding: 40px;
  background-color: white;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);

  h1 {
    font-size: 24px;
    margin: 0 0 20px 0;
    text-align: center;
  }

  p {
    margin-bottom: 20px;
    text-align: center;
  }

  footer {
    text-align: center;
  }
`;
