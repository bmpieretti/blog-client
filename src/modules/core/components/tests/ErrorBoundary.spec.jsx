import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson, mountToJson } from 'enzyme-to-json';
import faker from 'faker';
import ErrorBoundary from '../ErrorBoundary';

describe('ErrorBoundary component: ', () => {
  test('should render when error is caught', () => {
    const error = {
      message: 'message'
    };

    const info = {
      componentStack: 'info'
    };

    const component = mount(<ErrorBoundary />);

    // Calling componentDidCatch directly seems to only return the shallowWrapper instead,
    // seems to be an issue on Ezymes side due to an open issue on github
    component.setState({
      error: error.message,
      infoStack: info.componentStack
    });

    expect(mountToJson(component)).toMatchSnapshot();
    component.unmount();
  });

  test('should send email with passed error and info stack', () => {
    const error = {
      message: faker.lorem.word()
    };

    const info = {
      componentStack: faker.lorem.word()
    };

    const emailValue = 'mailto:bmpieretti@gmail.com';
    const subjectLine = '?subject=Incident%20Report&amp;';
    const body = `body=${error.message}--${info.componentStack}`;

    const component = mount(<ErrorBoundary />);
    component.setState({
      error: error.message,
      infoStack: info.componentStack
    });

    expect(component.find('a').getDOMNode().href).toEqual(`${emailValue}${subjectLine}${body}`);
    component.unmount();
  });

  test('should render children when no error', () => {
    const unrenderedElement = (
      <ErrorBoundary>
        <div />
      </ErrorBoundary>
    );

    const component = shallow(unrenderedElement);
    expect(shallowToJson(component)).toMatchSnapshot();
  });
});
