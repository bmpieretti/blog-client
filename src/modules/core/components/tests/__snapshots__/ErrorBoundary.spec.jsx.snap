// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary component:  should render children when no error 1`] = `<div />`;

exports[`ErrorBoundary component:  should render when error is caught 1`] = `
.c0 {
  max-width: 480px;
  margin: 0 auto;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  padding: 40px;
  background-color: white;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
}

.c0 h1 {
  font-size: 24px;
  margin: 0 0 20px 0;
  text-align: center;
}

.c0 p {
  margin-bottom: 20px;
  text-align: center;
}

.c0 footer {
  text-align: center;
}

.c1 {
  color: #f33;
}

<ErrorBoundary>
  <ErrorBoundary___default>
    <div
      className="c0"
    >
      <h1>
        Oh no! You have hit an error that should never have made it into the app!
      </h1>
      <Base__ErrorP>
        <p
          className="c1"
        >
          message
        </p>
      </Base__ErrorP>
      <footer>
        Let me know about this problem via email at 
        <a
          href="mailto:bmpieretti@gmail.com?subject=Incident%20Report&amp;body=message--info"
        >
          bmpieretti@gmail.com
        </a>
        .
      </footer>
    </div>
  </ErrorBoundary___default>
</ErrorBoundary>
`;
